File: Toocal.Core/Node.fs
module Toocal.Core.Node

open System
open DataAccessLayer
open Page

type Item =
  struct
    val Key : Byte[]
    val Value : Byte[]

    new (key, value) = { Key = key ; Value = value }
  end

type DataAccessLayer with

  member public this.GetNode (pageNum : PageNum) =
    let node = Node this

    task {
      let! page = this.ReadPage pageNum
      node.Deserialize page.Data
      node.SetPageNum page.Num
      return node
    }

  member public this.WriteNode (node : Node) =
    let page = this.AllocateEmptyPage ()

    if node.PageNum = 0UL then
      page.SetNum (this.FreeList.NextPage ())
      node.SetPageNum page.Num
    else
      page.SetNum node.PageNum

    node.Serialize page.Data

    task {
      do! this.WritePage page
      return node
    }

  member inline public this.DeleteNode (pageNum : PageNum) = this.FreeList.ReleasePage pageNum

and Node =
  struct
    val mutable PageNum : PageNum
    val Items : Collections.Generic.List<Item>
    val Children : Collections.Generic.List<PageNum>
    val Dal : DataAccessLayer

    new (dal : DataAccessLayer) =
      {
        PageNum = 0UL
        Items = Collections.Generic.List<Item> ()
        Children = Collections.Generic.List<PageNum> ()
        Dal = dal
      }

    member public this.FindWithKey (key : Byte[]) =
      let mutable compare = false

      let index =
        this.Items.FindIndex (fun item ->
          match Array.compareWith (fun (k1 : Byte) k2 -> k1.CompareTo k2) item.Key key with
          | 0 ->
            compare <- true
            true
          | 1 ->
            compare <- false
            true
          | _ -> false)

      if index = -1 then false, this.Items.Count else compare, index

    member public this.FindKey (key : Byte[]) =
      // Byrefs cannot be captured by closures or passed to inner functions.
      let this = this
      task { return! Node.FindKeyHelper (this, key) }

    static member private FindKeyHelper (node : Node, key : Byte[]) =
      let wasFound, index = node.FindWithKey key

      if wasFound then
        task { return index, Some node }
      elif node.IsLeaf () then
        task { return -1, None }
      else
        task {
          let! node = node.Dal.GetNode (node.Children[index])
          return! Node.FindKeyHelper (node, key)
        }

    member inline public this.WriteNodes (nodes : Node[]) =
      let resultArray = Array.zeroCreate<Threading.Tasks.Task<Node>> nodes.Length

      for index = 0 to (resultArray.Length - 1) do
        resultArray[index] <- this.Dal.WriteNode (nodes[index])

      resultArray

    member public this.SetPageNum (pageNum : PageNum) = this.PageNum <- pageNum

    member private this.IsLeaf () = this.Children.Count = 0

    member public this.Serialize (buffer : Byte[]) =
      this.SerializeHeader (buffer, this.IsLeaf (), 0, buffer.Length - 1) |> this.SerializeItems

    member public this.Deserialize (buffer : Byte[]) = this.DeserializeHeader (buffer, 0) |> this.DeserializeItems

    member private this.DeserializeHeader (buffer : Byte[], left : int) =
      let mutable left = left
      let isLeaf = buffer[0] |> uint16
      let itemsCount = buffer[1..2] |> BitConverter.ToInt32
      left <- left + 3
      (buffer, isLeaf, itemsCount, left)

    member private this.DeserializeItems (buffer : Byte[], isLeaf : uint16, itemsCount : int, left : int) =
      let mutable left = left

      for i = 0 to itemsCount - 1 do
        if isLeaf = 0us then
          this.Children.Add (buffer[left..] |> BitConverter.ToUInt64)
          left <- left + Page.SIZE

        let mutable offset = buffer[left..] |> BitConverter.ToUInt16
        left <- left + 2

        let keyLen = buffer[int offset] |> uint16
        offset <- offset + 1us
        let key = buffer[int offset .. int (offset + keyLen - 1us)]
        offset <- offset + keyLen

        let valueLen = buffer[int offset] |> uint16
        offset <- offset + 1us
        let value = buffer[int offset .. int (offset + valueLen - 1us)]
        offset <- offset + valueLen

        this.Items.Add (Item (key, value))

      if isLeaf = 0us then this.Children.Add (buffer[left..] |> BitConverter.ToUInt64)

    member private this.SerializeItems (buffer : Byte[], isLeaf : bool, left : int, right : int) =
      let mutable left = left
      let mutable right = right

      for i = 0 to this.Items.Count - 1 do
        let item = this.Items[i]

        if not isLeaf then
          let child = this.Children[i] |> BitConverter.GetBytes
          Array.blit child 0 buffer left child.Length
          left <- Page.SIZE

        let offset = (right - item.Key.Length - item.Value.Length - 2) |> uint16 |> BitConverter.GetBytes

        Array.blit offset 0 buffer left offset.Length
        left <- left + 2

        right <- right - item.Value.Length
        Array.blit item.Value 0 buffer right item.Value.Length
        right <- right - 1
        buffer[right] <- item.Value.Length |> Convert.ToByte

        right <- right - item.Key.Length
        Array.blit item.Key 0 buffer right item.Key.Length
        right <- right - 1
        buffer[right] <- item.Key.Length |> Convert.ToByte

    member private this.SerializeHeader (buffer : Byte[], isLeaf : bool, left : int, right : int) =
      let mutable left = left
      let mutable right = right

      buffer[left] <- if isLeaf then 1uy else 0uy
      left <- left + 1

      // Add the Key-Value pair count
      let keyValuePairCount = this.Items.Count |> uint16 |> BitConverter.GetBytes

      Array.blit keyValuePairCount 0 buffer left keyValuePairCount.Length
      left <- left + 2

      (buffer, isLeaf, left, right)

  end

File: src/main/scala/com/muqiuhan/toocal/core/Node.scala
package com.muqiuhan.toocal.core

import java.nio.ByteBuffer
import com.muqiuhan.toocal.errors.Error

import java.util
import scala.compiletime.ops.boolean
import scala.util.boundary
import boundary.break
import scala.annotation.tailrec

sealed case class Item(
    key: Array[Byte],
    value: Array[Byte]
):
    inline def size(): Int = key.length + value.length + Page.SIZE
end Item

class Node(dal: DataAccessLayer):
    var pageNumber: PageNumber = -1
    var items                  = new collection.mutable.ArrayBuffer[Item]()
    var children               = new collection.mutable.ArrayBuffer[PageNumber]()

    inline def isLeaf: Boolean = children.isEmpty

    inline def getNode(pageNumber: PageNumber): Node = dal.getNode(pageNumber).fold(_.raise(), identity)
    inline def writeNode(node: Node): Node           = dal.writeNode(node).fold(_.raise(), identity)
    inline def writeNodes(nodes: Node*): Unit        = nodes.foreach(writeNode)
    inline def indexItemSize(index: Int): Int        = items(index).key.length + items(index).value.length + Page.SIZE
    inline def isOverPopulated: Boolean              = dal.isOverPopulated(this)
    inline def isUnderPopulated: Boolean             = dal.isUnderPopulated(this)

    private def _findWithKey(key: Array[Byte]): (Boolean, Int) =
        boundary:
            for i <- items.indices do
                util.Arrays.compare(items(i).key, key) match
                    case 0 => break((true, i))
                    case 1 => break((false, i))
                    case _ => ()
            end for
            (false, items.length)
    end _findWithKey

    inline def findWithKey(key: Array[Byte]): Either[Error, (Int, Node)] =
        Node.findKey(this, key) match
            case (_, None)           => Left(Error.FindNodeWithKeyError(key))
            case (index, Some(node)) => Right((index, node))
        end match
    end findWithKey

    inline def nodeSize: Int =
        items.foldLeft(Node.HEADER_SIZE)((size: Int, item: Item) =>
            size + item.size()
        ) + Page.SIZE

    inline def addItem(item: Item, insertionIndex: Int): Int =
        items.insert(insertionIndex, item)
        insertionIndex

    def serialize(buffer: ByteBuffer): Unit =
        val isLeafFlag = isLeaf

        buffer.put((if isLeafFlag then 1 else 0).toByte)
        buffer.putInt(items.length)

        for i <- items.indices do
            val item = items(i)
            if !isLeafFlag then buffer.putLong(children(i))

            buffer
                .putInt(item.key.length)
                .putInt(item.value.length)
                .put(item.key)
                .put(item.value)
        end for

        if !isLeafFlag then buffer.putLong(children.last)
    end serialize

    def deserialize(buffer: ByteBuffer): Unit =
        val isLeafFlag = buffer.get() == 0

        for i <- 0 until buffer.getInt() do
            if !isLeafFlag then children.addOne(buffer.getLong())

            val key   = new Array[Byte](buffer.getInt())
            val value = new Array[Byte](buffer.getInt())

            buffer.get(key).get(value)
            items.addOne(Item(key, value))
        end for
    end deserialize
end Node

object Node:
    val HEADER_SIZE: Int = 3

    @tailrec
    def findKey(node: Node, key: Array[Byte]): (Int, Option[Node]) =
        val (wasFound, index) = node._findWithKey(key)

        if wasFound then return (index, Some(node))
        if node.isLeaf then return (-1, None)

        findKey(node.getNode(node.children(index)), key)
    end findKey
end Node

extension (self: DataAccessLayer)
    /** Get the corresponding page and deserialize it to node.
      * 
      * @param pageNumber The page number corresponding to Node.
      * @return If failure returns Error.DataAccessLayerCannotGetNode
      */
    def getNode(pageNumber: PageNumber): Either[Error, Node] =
        self.operatingAtPage(
            pageNumber,
            () =>
                self.readPage(pageNumber).flatMap(page =>
                    val node = new Node(self)
                    node.deserialize(page.data)
                    node.pageNumber = pageNumber
                    Right(node)
                ).orElse(Left(Error.DataAccessLayerCannotGetNode))
        )
    end getNode

    /** Serialize and write the node to the corresponding page.
      * 
      * @param node The page number corresponding to Node.
      * @return If failure returns Error.DataAccessLayerCannotWriteNode
      */
    def writeNode(node: Node): Either[Error, Node] =
        if node.pageNumber == 0 then
            node.pageNumber = self.freelist.getNextPage

        val page = self.allocateEmptyPage(node.pageNumber)
        node.serialize(page.data)

        self.operatingAtPage(
            self.meta.freelistPage,
            () =>
                self.writePage(page) match
                    case Left(_)      => Left(Error.DataAccessLayerCannotWriteNode)
                    case Right(value) => Right(node)
        )
    end writeNode

    def getSplitIndex(node: Node): Int =
        var size = Node.HEADER_SIZE

        boundary:
            for i <- node.items.indices do
                size += node.indexItemSize(i)
                if size > self.minThreshold && i < node.items.length - 1 then
                    break(i + 1)
            end for

        -1
    end getSplitIndex

    /** Delete a Node and release the page where it is located directly from the freelist.
      * 
      * @param pageNumber The page number of node.
      */
    inline def deleteNode(pageNumber: PageNumber): Unit = self.freelist.releasePage(pageNumber)
    inline def isOverPopulated(node: Node): Boolean  = node.nodeSize > self.maxThreshold
    inline def isUnderPopulated(node: Node): Boolean = node.nodeSize < self.maxThreshold
end extension
